### PR TITLE
Add support for gobject closures

### PR DESCRIPTION
**Pull request by antoyo**
Hi.
I started to create the `Closure` type so that we can generate the code of methods taking a `GClosure` as a parameter.
This is a work in progress, but I open this pull request right now, which is working (even thought it contains a hack) so that you can comment on it.
I'll have to fix the hack for `size_of` by writing a C function to get the size of `GClosure`, unless someone has a better idea.

I think that marshallers support should be done in another pull request since I believe that doing something both safe and efficient will be non trivial.

### CODE REVIEW

File: src/closure.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+use Value;
+use ToValue;
+
+pub struct Closure {
+    closure: *mut gobject_ffi::GClosure,
+}
+
+impl Closure {
+    pub fn new<F: Fn(&[Value]) -> Option<Value> + 'static>(callback: F) -> Self {
+        unsafe extern "C" fn marshal<F: Fn(&[Value]) -> Option<Value> + 'static>(closure: *mut gobject_ffi::GClosure,
```
Comment (EPashkin) [inline src/closure.rs]: Rename parameter to `_closure` if it stay unused and `_invocation_hint` too

File: src/closure.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            }
+            for value in values {
+                mem::forget(value);
+            }
+        }
+
+        unsafe {
+            // TODO: get the appropriate size with size_of (not working since the generated
+            // GClosure does not contain every field).
+            let closure = gobject_ffi::g_closure_new_simple(1000, ptr::null_mut());
```
Comment (EPashkin) [inline src/closure.rs]: IMHO `g_closure_new` need used with something like https://github.com/gtk-rs/glib/blob/master/src/signal.rs#L64-L68

**Comment by EPashkin** [comment]
Just added minor tips, as I don't fully understand closures and don't sure how to test it in wild ;)

File: src/value.rs [hunk 3]
```diff
@@ -215,6 +217,40 @@ impl Uninitialized for Value {
     }
 }
 
+impl FromGlibPtrNone<*mut gobject_ffi::GValue> for Value {
+    #[inline]
+    unsafe fn from_glib_none(ptr: *mut gobject_ffi::GValue) -> Self {
+        assert!(!ptr.is_null());
+        Value(ptr::read(ptr))
```
Comment (sdroege) [inline src/value.rs]: This causes two owners of the same memory if the GValue is some kind of pointer (object, string, boxed type, ...). Both your new Value and the C GValue will contain the same pointer, and both think they own it and will try to free it later, leaving the other one with a pointer to a now invalid memory address.

File: src/value.rs [hunk 4]
```diff
@@ -215,6 +217,40 @@ impl Uninitialized for Value {
...
+    }
+}
+
+impl FromGlibPtrFull<*mut gobject_ffi::GValue> for Value {
+    #[inline]
+    unsafe fn from_glib_full(ptr: *mut gobject_ffi::GValue) -> Self {
+        let value = Value(ptr::read(ptr));
+        gobject_ffi::g_value_unset(ptr);
+        value
```
Comment (sdroege) [inline src/value.rs]: This one doesn't have the above problem, because the C code passed you ownership of whatever the GValue contains, but it also passed you ownership of the GValue itself. You need to free it afterwards, or leak memory.

File: src/value.rs [hunk 5]
```diff
@@ -215,6 +217,40 @@ impl Uninitialized for Value {
...
+    #[inline]
+    unsafe fn from_glib_none(ptr: *const gobject_ffi::GValue) -> Self {
+        assert!(!ptr.is_null());
+        Value(ptr::read(ptr))
+    }
+}
+
+impl FromGlibPtrFull<*const gobject_ffi::GValue> for Value {
+    #[inline]
+    unsafe fn from_glib_full(ptr: *const gobject_ffi::GValue) -> Self {
```
Comment (sdroege) [inline src/value.rs]: A `*const GValue` seems wrong here. C code is never going to mark a pointer a `const` if you're going to own it, as otherwise all mutating functions (and free(), g_value_unset(), etc) would cause a compiler warning (you can't pass const pointers to them).

File: src/value.rs (hunk 4 above)
Comment (sdroege) [inline src/value.rs]: Also it has actually the same problem, just slightly different. If you didn't have the g_value_unset(), it would be fine. But because of the g_value_unset(), for things stored as a pointer you will copy over the pointer and then immediately free it, leaving you with a pointer to a now invalid memory address.

File: src/closure.rs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                values.push(value);
+                param_values = param_values.offset(1);
+            }
+            let callback: Box<Box<F>> = Box::from_raw(marshal_data as *mut _);
+            let result = callback(&values);
+            if let Some(result) = result {
+                *return_value = result.into_raw();
+            }
+            for value in values {
+                mem::forget(value);
```
Comment (sdroege) [inline src/closure.rs]: And this is why the FromGlib* impls work for you. Once fixed, you would have to remove the mem::forget() here and let the Drop impls take care of everything else

File: src/closure.rs (hunk 2 above)
Comment (sdroege) [inline src/closure.rs]: No, that's fine as is. But the size of GClosure is 4 + 3 * sizeof(void*)

File: src/closure.rs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+            .map(|v| v.to_value())
+            .collect();
+        let mut gvalues: Vec<gobject_ffi::GValue> = values.iter()
+            .map(|v| unsafe { ptr::read(v.to_glib_none().0) })
+            .collect();
+        unsafe {
+            gobject_ffi::g_closure_invoke(self.to_glib_none_mut().0, value.to_glib_none_mut().0, gvalues.len() as u32,
+                gvalues.as_mut_ptr(), ptr::null_mut());
+        }
+        Some(value)
```
Comment (sdroege) [inline src/closure.rs]: This might be unsafe. I think you should a) initialize value with mem::zeroed(), and b) check here if value.g_type == G_TYPE_NONE and in that case return None. What you do here can AFAIU return a Value full of uninitialized memory.

File: src/closure.rs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    }
+}
+
+impl Drop for Closure {
```
Comment (sdroege) [inline src/closure.rs]: You can (and probably should) also implement Clone around g_closure_ref() here

File: src/closure.rs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+        unsafe {
+            // TODO: get the appropriate size with size_of (not working since the generated
+            // GClosure does not contain every field).
+            let closure = gobject_ffi::g_closure_new_simple(1000, ptr::null_mut());
+            assert_ne!(closure, ptr::null_mut());
+            let callback = Box::new(Box::new(callback));
+            let ptr: *mut Box<F> = Box::into_raw(callback);
+            let ptr: *mut c_void = ptr as *mut _;
+            gobject_ffi::g_closure_set_meta_marshal(closure, ptr, Some(marshal::<F>));
```
Comment (sdroege) [inline src/closure.rs]: This will either leak the callback Box, or free it multiple times. g_closure_invoke() can be called multiple times.
What you could do is to install a finalize notifier (g_closure_add_finalize_notifier()) with the box, and destroy the box in there. And inside the marshaller you would always "forget" the box again at the end so it does not get destroyed.

Comment (sdroege) [inline src/closure.rs]: FWIW, your implementation frees it multiple times and uses the freed box

**Comment by antoyo** [comment]
I think I fixed all mentionned issues (thanks for your review).
I'll use the work of pull request #172 when it's merged.

**Comment by antoyo** [comment]
I made a quick test with @sdroege implementation and my closure implementation with `webkit2gtk-webextension` and it does not seem to work well:
the first time, the callback is called, but then, the webview freeze and obviously, the callback is not called anymore after that.
Perhaps it would be easier to debug with a function from gobject/glib.

File: src/closure.rs [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+            gobject_ffi::g_closure_add_finalize_notifier(closure, ptr, Some(finalize::<F>));
+            Closure {
+                closure: closure,
+            }
+        }
+    }
+
+    pub fn invoke(&mut self, values: &[&ToValue]) -> Option<Value> {
```
Comment (sdroege) [inline src/closure.rs]: This does not need &mut btw. It's safe (as long as your closure is!) to call invoke() from multiple threads at once

File: src/closure.rs [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        let mut result: gobject_ffi::GValue = unsafe { mem::zeroed() };
+        let values: Vec<_> = values.iter()
+            .map(|v| v.to_value())
+            .collect();
+        let mut gvalues: Vec<gobject_ffi::GValue> = values.iter()
+            .map(|v| unsafe { ptr::read(v.to_glib_none().0) })
```
Comment (sdroege) [inline src/closure.rs]: To prevent copying, here it would be safe to just ptr::read(&v.0)

File: src/closure.rs [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+impl Closure {
+    pub fn new<F: Fn(&[Value]) -> Option<Value> + 'static>(callback: F) -> Self {
+        unsafe extern "C" fn marshal<F: Fn(&[Value]) -> Option<Value> + 'static>(_closure: *mut gobject_ffi::GClosure,
+            return_value: *mut gobject_ffi::GValue, n_param_values: c_uint, mut param_values: *mut gobject_ffi::GValue,
+            _invocation_hint: *mut c_void, marshal_data: *mut c_void)
+        {
+            let mut values = vec![];
+            for _ in 0 .. n_param_values {
+                let value: Value = FromGlibPtrNone::from_glib_none(param_values);
```
Comment (sdroege) [inline src/closure.rs]: And here to prevent copying, you could convert the `*mut GValue` to a `&Value` via pointer casting and dereference

File: src/closure.rs [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+        {
+            let _callback: Box<Box<F>> = Box::from_raw(notify_data as *mut _);
+            // callback is dropped here.
+        }
+
+        unsafe {
+            let size = 4 + 4 + 3 * mem::size_of::<*mut c_void>() as u32;
+            let closure = gobject_ffi::g_closure_new_simple(size, ptr::null_mut());
+            assert_ne!(closure, ptr::null_mut());
+            let callback = Box::new(Box::new(callback));
```
Comment (sdroege) [inline src/closure.rs]: Why do you Box twice btw?

File: src/closure.rs [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+            return_value: *mut gobject_ffi::GValue, n_param_values: c_uint, mut param_values: *mut gobject_ffi::GValue,
+            _invocation_hint: *mut c_void, marshal_data: *mut c_void)
+        {
+            let mut values = vec![];
+            for _ in 0 .. n_param_values {
+                let value: Value = FromGlibPtrNone::from_glib_none(param_values);
+                values.push(value);
+                param_values = param_values.offset(1);
+            }
+            let callback: &Box<F> = mem::transmute(marshal_data as *mut _);
```
Comment (sdroege) [inline src/closure.rs]: You can also use Box::from_raw() here and just mem::forget() it before returning.

You should also create a glib::CallbackGuard() here to catch panics and exit on panics

File: src/closure.rs [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+            let result = callback(&values);
+            match result {
+                Some(result) => *return_value = result.into_raw(),
+                None => {
+                    let result = Value::uninitialized();
+                    let mut gvalue = result.into_raw();
+                    gvalue.g_type = gobject_ffi::G_TYPE_NONE;
```
Comment (sdroege) [inline src/closure.rs]: Value::uninitialized() already sets the type to G_TYPE_NONE

**Comment by antoyo** [comment]
@sdroege You can test the `main` example in [the webkit2gtk-rs repository](https://github.com/antoyo/webkit2gtk-rs) with `cargo run --example --features v2_6`.
It requires that the example in [this repository](https://github.com/antoyo/webkit2gtk-webextension-rs/tree/feature/closure) to be built in debug mode (be sure to put both repositories at the same directory level so that [this path](https://github.com/antoyo/webkit2gtk-rs/blob/master/examples/main.rs#L36) works.
This will make the webview freeze after you clicked on the page: no more scroll or click working after that.

Since I used your `ToGlibPtr` implementations, this freeze cannot be caused by my implementation being unsound.

**Comment by sdroege** [comment]
Seems to work fine for me, no freezes or anything

File: src/closure.rs [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+
+impl<'a> ToGlibPtr<'a, *const gobject_ffi::GClosure> for Closure {
+    type Storage = &'a Closure;
+
+    fn to_glib_none(&'a self) -> Stash<'a, *const gobject_ffi::GClosure, Self> {
+        Stash(self.closure, self)
+    }
+}
+
+impl<'a> ToGlibPtrMut<'a, *mut gobject_ffi::GClosure> for Closure {
```
Comment (sdroege) [inline src/closure.rs]: You probably also want to implement the FromGlibPtr* traits

File: src/closure.rs [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+            .collect();
+        unsafe {
+            gobject_ffi::g_closure_invoke(self.to_glib_none_mut().0, &mut result, gvalues.len() as u32,
+                gvalues.as_mut_ptr(), ptr::null_mut());
+        }
+        if result.g_type == gobject_ffi::G_TYPE_NONE {
+            None
+        }
+        else {
+            Some(unsafe { Value::from_glib_full(&mut result as *mut _) })
```
Comment (sdroege) [inline src/closure.rs]: This should work:

let mut result = Value::uninitialized();
...
gobject_ffi::g_closure_invoke(..., result.to_glib_none_mut().0, ...)
...
Some(value)

File: src/closure.rs [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+            .map(|v| v.to_value())
+            .collect();
+        let mut gvalues: Vec<gobject_ffi::GValue> = values.iter()
+            .map(|v| unsafe { ptr::read(v.to_glib_none().0) })
+            .collect();
+        unsafe {
+            gobject_ffi::g_closure_invoke(self.to_glib_none_mut().0, &mut result, gvalues.len() as u32,
+                gvalues.as_mut_ptr(), ptr::null_mut());
+        }
+        if result.g_type == gobject_ffi::G_TYPE_NONE {
```
Comment (sdroege) [inline src/closure.rs]: G_TYPE_INVALID, not G_TYPE_NONE

File: src/closure.rs (hunk 17 above)
Comment (sdroege) [inline src/closure.rs]: Or alternatively something like `into_raw` in reverse (`from_raw`)

File: src/closure.rs [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+            }
+            let callback: &Box<F> = mem::transmute(marshal_data as *mut _);
+            let result = callback(&values);
+            match result {
+                Some(result) => *return_value = result.into_raw(),
+                None => {
+                    let result = Value::uninitialized();
+                    let mut gvalue = result.into_raw();
+                    gvalue.g_type = gobject_ffi::G_TYPE_NONE;
+                    *return_value = gvalue;
```
Comment (sdroege) [inline src/closure.rs]: You also want to check first `if !return_value.is_null()`

File: src/closure.rs [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+        unsafe extern "C" fn finalize<F: Fn(&[Value]) -> Option<Value> + 'static>(notify_data: *mut c_void,
+            _closure: *mut gobject_ffi::GClosure)
+        {
+            let _callback: Box<Box<F>> = Box::from_raw(notify_data as *mut _);
+            // callback is dropped here.
+        }
+
+        unsafe {
+            let size = 4 + 4 + 3 * mem::size_of::<*mut c_void>() as u32;
+            let closure = gobject_ffi::g_closure_new_simple(size, ptr::null_mut());
```
Comment (sdroege) [inline src/closure.rs]: And you want to call g_closure_sink() after this. Floating references \o/  (fixes crashes, testcase coming in a minute)

**Comment by sdroege** [comment]
Standalone testcase here, with fixed up (hopefully) Closure bindings: https://gist.github.com/sdroege/d47508b39f55bf454f90673150ccef20
Is still missing the optimizations to prevent unneeded copying though

File: src/closure.rs (hunk 17 above)
Comment (EPashkin) [inline src/closure.rs]: Note: Instead `result.to_glib_none_mut().0`  better use `mut_override(result.to_glib_none().0)` in most cases

File: src/closure.rs (hunk 14 above)
Comment (antoyo) [inline src/closure.rs]: Isn't it better to transmute to a reference to avoid having to `mem::forget`?
I believe both options has some drawbacks.

File: src/closure.rs (hunk 13 above)
Comment (antoyo) [inline src/closure.rs]: I followed the style used by the code generated by `gir`, for instance in [`gtk::button`](https://github.com/gtk-rs/gtk/blob/master/src/auto/button.rs#L294).
I don't know the reason to double box: can @GuillaumeGomez or @EPashkin explain, please?

Comment (EPashkin) [inline src/closure.rs]: I don't remember clearly :(
Maybe for "safety" if outer box restored to broken object we get AV when try deref inner

**Comment by sdroege** [comment]
Maybe use the glib_wrapper!() macro here with the ref/unref functions, it would automatically give you most of the trait impls. And instead of calling g_closure_ref() & g_closure_sink only in the constructor, the best would be to make that the ref function. I.e.
```Rust
glib_wrapper! {
     pub struct Closure(Shared<ffi::Closure>);

     match fn {
         ref => |ptr| {
             g_closure_ref(ptr);
             g_closure_sink(ptr);
         },
         unref => |ptr| ffi::g_closure_unref(ptr),
     }
 }
```

File: src/closure.rs [hunk 21]
```diff
@@ -23,31 +23,31 @@ impl Closure {
             _invocation_hint: *mut c_void, marshal_data: *mut c_void)
         {
             let values = slice::from_raw_parts(param_values as *const _, n_param_values as usize);
-            let callback: &Box<F> = mem::transmute(marshal_data as *mut _);
+            let callback: Box<F> = Box::from_raw(marshal_data as *mut _);
```
Comment (sdroege) [inline src/closure.rs]: If you want to omit the mem::forget() below, you could also do `let callback: &F = mem::transmute(marshal_data)`

**Comment by antoyo** [comment]
I've just tested the webkit example with the latest version of this pull request and it now works fine.

File: Cargo.toml [hunk 22]
```diff
@@ -24,7 +24,7 @@ version = "0.3.4"
 
 [dependencies.gobject-sys]
 git = "https://github.com/gtk-rs/sys"
-version = "0.3.4"
+#version = "0.3.4"
```
Comment (EPashkin) [inline Cargo.toml]: Forgot to remove temporary change?

File: src/closure.rs [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+
+use gobject_ffi;
+use translate::{Stash, StashMut, ToGlibPtr, ToGlibPtrMut, Uninitialized, mut_override};
+use types::Type;
+use Value;
+use ToValue;
+
+pub struct Closure {
+    closure: *mut gobject_ffi::GClosure,
+}
```
Comment (sdroege) [inline src/closure.rs]: glib_wrapper! {
     pub struct Closure(Shared<ffi::Closure>);

     match fn {
         ref => |ptr| {
             g_closure_ref(ptr);
             g_closure_sink(ptr);
         },
         unref => |ptr| ffi::g_closure_unref(ptr),
     }
 }

File: src/closure.rs [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+            let callback = Box::new(callback);
+            let ptr: *mut F = Box::into_raw(callback);
+            let ptr: *mut c_void = ptr as *mut _;
+            gobject_ffi::g_closure_set_meta_marshal(closure, ptr, Some(marshal::<F>));
+            gobject_ffi::g_closure_add_finalize_notifier(closure, ptr, Some(finalize::<F>));
+            gobject_ffi::g_closure_ref(closure);
+            gobject_ffi::g_closure_sink(closure);
+            Closure {
+                closure: closure,
+            }
```
Comment (sdroege) [inline src/closure.rs]: from_glib_none(closure) (and remove the manual ref/sink calls)

File: src/closure.rs [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+        }
+    }
+
+    pub fn invoke(&self, values: &[&ToValue]) -> Option<Value> {
+        let result = unsafe { Value::uninitialized() };
+        let mut values: Vec<_> = values.iter()
+            .map(|v| v.to_value())
+            .collect();
```
Comment (sdroege) [inline src/closure.rs]: If we would take &[Value] here, we could also prevent the last copy

File: src/closure.rs [hunk 26]
```diff
@@ -0,0 +1,157 @@
...
+        let gvalues = values.as_mut_ptr() as *mut _;
+        unsafe {
+            gobject_ffi::g_closure_invoke(self.to_glib_none().0 as *mut _, mut_override(result.to_glib_none().0),
```
Comment (sdroege) [inline src/closure.rs]: Why mut_override() instead of result.to_glib_none_mut()?

File: src/closure.rs (hunk 16 above)
Comment (sdroege) [inline src/closure.rs]: Drop, Clone, ToGlib*, FromGlib* would all be automatically implemented by using `glib_wrapper!`

**Comment by sdroege** [comment]
Generally looks good to me now, just some further comments

File: src/closure.rs (hunk 26 above)
Comment (EPashkin) [inline src/closure.rs]: Agree, mut_override don't need here as result is output parameter

File: src/closure.rs [hunk 27]
```diff
@@ -0,0 +1,134 @@
+// TODO: support marshaller.
```
Comment (EPashkin) [inline src/closure.rs]: Please, don't forget add standard header

File: src/closure.rs (hunk 25 above)
Comment (antoyo) [inline src/closure.rs]: @sdroege: You mean that the parametr `values` should be `&[Value]` instead of `&[&ToValue]`?

File: src/closure.rs (hunk 20 above)
Comment (antoyo) [inline src/closure.rs]: I think I need to remove these calls, now that I use `glib_wrapper`, right?

File: src/closure.rs [hunk 28]
```diff
@@ -0,0 +1,138 @@
...
+            let size = 4 + 4 + 3 * mem::size_of::<*mut c_void>() as u32;
+            let closure = gobject_ffi::g_closure_new_simple(size, ptr::null_mut());
+            assert_ne!(closure, ptr::null_mut());
+            let callback = Box::new(callback);
+            let ptr: *mut F = Box::into_raw(callback);
+            let ptr: *mut c_void = ptr as *mut _;
+            gobject_ffi::g_closure_set_meta_marshal(closure, ptr, Some(marshal::<F>));
+            gobject_ffi::g_closure_add_finalize_notifier(closure, ptr, Some(finalize::<F>));
+            gobject_ffi::g_closure_ref(closure);
+            gobject_ffi::g_closure_sink(closure);
```
Comment (sdroege) [inline src/closure.rs]: You can remove the manual calls to `ref` and `sink` here now

File: src/closure.rs (hunk 20 above)
Comment (sdroege) [inline src/closure.rs]: Yes

File: src/closure.rs (hunk 25 above)
Comment (sdroege) [inline src/closure.rs]: Not sure, but if it was you wouldn't have to allocate the Vec and could do everything without any copying or allocations. But with `&[&ToValue]` it's nicer for the user to call it.

**Comment by sdroege** [comment]
No further comments from me 👍

**Comment by antoyo** [comment]
All the tests passed, is there any other comment you want to make?

File: src/closure.rs [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+            let ptr: *mut c_void = ptr as *mut _;
+            gobject_ffi::g_closure_set_meta_marshal(closure, ptr, Some(marshal::<F>));
+            gobject_ffi::g_closure_add_finalize_notifier(closure, ptr, Some(finalize::<F>));
+            from_glib_none(closure)
+        }
+    }
+
+    pub fn invoke(&self, values: &[&ToValue]) -> Option<Value> {
+        let mut result = unsafe { Value::uninitialized() };
+        let mut values: Vec<_> = values.iter()
```
Comment (GuillaumeGomez) [inline src/closure.rs]: Ii's more of a style issue, so not mandatory, but I'd prefer it to be like this:

```rust
let mut values: Vec<_> = values.iter()
                                .map(|v| v.to_value())
                                .collect();
```

File: src/closure.rs [hunk 30]
```diff
@@ -0,0 +1,136 @@
...
+            .collect();
+        let gvalues = values.as_mut_ptr() as *mut _;
+        unsafe {
+            gobject_ffi::g_closure_invoke(self.to_glib_none().0 as *mut _, result.to_glib_none_mut().0,
+                values.len() as u32, gvalues, ptr::null_mut());
+        }
+        if result.type_() == Type::Invalid {
+            None
+        }
+        else {
```
Comment (GuillaumeGomez) [inline src/closure.rs]: Still a style issue (so still not mandatory): could you put the `else {` on the same line as `}` please?

**Comment by GuillaumeGomez** [comment]
I made some but otherwise seems all good to me.

File: src/closure.rs [hunk 31]
```diff
@@ -0,0 +1,136 @@
...
+        {
+            let values = slice::from_raw_parts(param_values as *const _, n_param_values as usize);
+            let callback: Box<F> = Box::from_raw(marshal_data as *mut _);
+            let result = callback(values);
+            if !return_value.is_null() {
+                match result {
+                    Some(result) => *return_value = result.into_raw(),
+                    None => {
+                        let result = Value::uninitialized();
+                        *return_value = result.into_raw();
```
Comment (EPashkin) [inline src/closure.rs]: Why don't call g_value_init(return_value)?

Comment (sdroege) [inline src/closure.rs]: There's no type it could be initialized with here. We just need to fill the return value with something here, just in case. And that something is a GValue initialized with all zeroes (i.e. G_TYPE_INVALID), something we can detect later and handle appropriately.

Comment (EPashkin) [inline src/closure.rs]: You right, Type::Invalid don't accepted by g_value_init

Comment (sdroege) [inline src/closure.rs]: Value::uninitialized() is a bit misleading IMHO, it's not uninitialized but all-zeroes :)

File: src/closure.rs (hunk 29 above)
Comment (antoyo) [inline src/closure.rs]: I believe it is best to be consistent, whatever style we choose.
For instance, I used the same style as [this](https://github.com/gtk-rs/glib/blob/master/src/value.rs#L238-L240).

File: src/closure.rs [hunk 32]
```diff
@@ -0,0 +1,136 @@
...
+        ref => |ptr| {
+            gobject_ffi::g_closure_ref(ptr);
+            gobject_ffi::g_closure_sink(ptr);
+        },
+        unref => |ptr| gobject_ffi::g_closure_unref(ptr),
+    }
+}
+
+impl Closure {
+    pub fn new<F: Fn(&[Value]) -> Option<Value> + 'static>(callback: F) -> Self {
```
Comment (sdroege) [inline src/closure.rs]: One problem I forgot about is that GClosure is `Sync+Send`. You have no guarantees from which thread your callback is called, you have no guarantees that it isn't called simultaneously from multiple threads at once. As such this should probably be `F: Fn(&[Value]) -> Option<Value> + Send + Sync + 'static`

File: src/closure.rs [hunk 33]
```diff
@@ -0,0 +1,136 @@
+// Copyright 2013-2017, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+// TODO: support marshaller.
```
Comment (sdroege) [inline src/closure.rs]: What does `TODO: support marshaller` mean? What do you want to support with marshallers?

Comment (antoyo) [inline src/closure.rs]: I would like to have something like `GCClosure`, i.e. being able to use a function with Rust types.
I believe it should be done in another pull request since it looks non-trivial to do something both safe and efficient.

Comment (sdroege) [inline src/closure.rs]: That would indeed be very interesting and is also on my list. For dynamically using signals/action signals
Separate pull request, yes

**Comment by GuillaumeGomez** [comment]
Ok, it seems good to me.

@EPashkin @sdroege: For you?

**Comment by EPashkin** [comment]
👍 Thanks, @antoyo 
And thanks @sdroege too for good ideas

**Comment by GuillaumeGomez** [comment]
Let's merge then. Thanks to everyone for your work in here! :)
